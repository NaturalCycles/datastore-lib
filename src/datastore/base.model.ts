import { idSchema, objectSchema, unixTimestampSchema, verSchema } from '@naturalcycles/nodejs-lib'

export enum ModelType {
  DBM = 'DBM',
  BM = 'BM',
  FM = 'FM',
}

export const baseDBEntitySchema = objectSchema({
  id: idSchema,
  created: unixTimestampSchema,
  updated: unixTimestampSchema,
  _ver: verSchema.optional(),
})
