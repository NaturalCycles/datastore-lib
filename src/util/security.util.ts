export type NanoidGenerate = (alphabet: string, length?: number) => string
const nanoidGenerate = require('nanoid/generate') as NanoidGenerate

export const ALPHABET_NUMBER = '0123456789'
export const ALPHABET_LOWERCASE = 'abcdefghijklmnopqrstuvwxyz'
export const ALPHABET_UPPERCASE = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
export const ALPHABET_ALPHANUMERIC_LOWERCASE = [ALPHABET_NUMBER, ALPHABET_LOWERCASE].join('')
export const ALPHABET_ALPHANUMERIC_UPPERCASE = [ALPHABET_NUMBER, ALPHABET_UPPERCASE].join('')
export const ALPHABET_ALPHANUMERIC = [ALPHABET_NUMBER, ALPHABET_LOWERCASE, ALPHABET_UPPERCASE].join(
  '',
)

export function generateStringId (length = 16, alphabet = ALPHABET_ALPHANUMERIC_LOWERCASE): string {
  return nanoidGenerate(alphabet, length)
}
