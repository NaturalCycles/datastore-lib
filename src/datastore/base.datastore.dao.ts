import { Query } from '@google-cloud/datastore'
import { ObjectSchema } from '@hapi/joi'
import { filterEmptyStringValues, objectNullValuesToUndefined } from '@naturalcycles/js-lib'
import { getValidationResult } from '@naturalcycles/nodejs-lib'
import { Observable } from 'rxjs'
import { flatMap, map } from 'rxjs/operators'
import { DBRelation, ModelType } from './base.model'
import { BaseDatastoreDaoCfg, DaoOptions, DatastoreKey } from './datastore.model'
import { DatastoreService } from './datastore.service'
import { createdUpdatedFields } from './model.util'

export abstract class BaseDatastoreDao<BM = any, DBM = BM, FM = BM> {
  constructor (
    private datastoreService: DatastoreService,
    private baseDatastoreDaoCfg: BaseDatastoreDaoCfg,
  ) {}

  abstract readonly KIND: string
  abstract excludeFromIndexes: string[] = []
  readonly BACKEND_RESPONSE_PROPERTY?: string
  readonly ACCOUNT_RELATION?: DBRelation
  readonly DBM_SCHEMA?: ObjectSchema
  readonly BM_SCHEMA?: ObjectSchema
  readonly FM_SCHEMA?: ObjectSchema

  // to be extended
  beforeCreate (bm: Partial<BM>): BM {
    return bm as BM
  }

  // to be extended
  beforeDBMValidate (dbm: DBM): DBM {
    return dbm
  }

  // to be extended
  async beforeDBMToBM (dbm: DBM): Promise<BM> {
    return dbm as any
  }

  // to be extended
  async beforeBMToDBM (bm: BM): Promise<DBM> {
    return bm as any
  }

  // to be extended
  async beforeBMToFM (bm: BM): Promise<FM> {
    return bm as any
  }

  // to be extended
  anonymize (dbm: DBM): DBM {
    return dbm
  }

  /**
   * Mutates object with properties: id, created, updated.
   * Returns DBM (new reference).
   */
  async bmToDBM (_bm: BM, opt: DaoOptions = {}): Promise<DBM> {
    if (_bm === undefined) return undefined as any

    let bm: BM = { ...(_bm as any) } // clone

    // Convert `null` values to `undefined` values after Java
    bm = objectNullValuesToUndefined(bm)
    bm = filterEmptyStringValues(bm)

    bm = this.datastoreService.assignIdCreatedUpdated(bm, opt.preserveUpdatedCreated)

    // Validate/convert BM
    // bm gets assigned to the new reference
    bm = this.validateAndConvert(bm, this.BM_SCHEMA, ModelType.BM, opt)

    // BM > DBM
    let dbm = await this.beforeBMToDBM(bm)

    // Validate/convert DBM
    // dbm gets assigned to the new reference
    dbm = this.validateAndConvert(dbm, this.DBM_SCHEMA, ModelType.DBM, opt)

    return dbm
  }

  async dbmsToBM (dbms: DBM[], opt: DaoOptions = {}): Promise<BM[]> {
    // try/catch if invalid?..
    return Promise.all(dbms.map(dbm => this.dbmToBM(dbm, opt)))
  }

  async bmsToDBM (bms: BM[], opt: DaoOptions = {}): Promise<DBM[]> {
    // try/catch?
    return Promise.all(bms.map(bm => this.bmToDBM(bm, opt)))
  }

  async bmsToFM (bms: BM[], opt: DaoOptions = {}): Promise<FM[]> {
    // try/catch?
    return Promise.all(bms.map(bm => this.bmToFM(bm, opt)))
  }

  anyArrayToDBM (entities: any[], opt: DaoOptions = {}): DBM[] {
    return entities.map(e => this.anyToDBM(e, opt))
  }

  anyToDBM (entity: any, opt: DaoOptions = {}): DBM {
    if (entity === undefined) return undefined as any
    // Convert `null` values to `undefined` values after Java
    let dbm: DBM = entity
    // Convert `null` values to `undefined` values after Java
    dbm = objectNullValuesToUndefined(dbm)
    dbm = filterEmptyStringValues(dbm)

    if (opt.anonymize) {
      dbm = this.anonymize(dbm)
    }

    // Validate/convert DBM
    // dbm gets assigned to the new reference
    dbm = this.validateAndConvert<DBM>(dbm, this.DBM_SCHEMA, ModelType.DBM, opt)

    return dbm
  }

  async dbmToBM (dbm: DBM, opt: DaoOptions = {}): Promise<BM> {
    if (dbm === undefined) return undefined as any

    dbm = this.anyToDBM(dbm, opt)

    // DBM > BM
    let bm = await this.beforeDBMToBM(dbm)

    // Validate/convert BM
    // bm gets assigned to the new reference
    bm = this.validateAndConvert<BM>(bm, this.BM_SCHEMA, ModelType.BM, opt)

    return bm
  }

  async bmToFM (bm: BM, opt: DaoOptions = {}): Promise<FM> {
    if (bm === undefined) return undefined as any

    // Validate/convert BM
    // bm gets assigned to the new reference
    bm = this.validateAndConvert<BM>(bm, this.BM_SCHEMA, ModelType.BM, opt)

    // BM > FM
    let fm = await this.beforeBMToFM(bm)

    // Validate/convert FM
    // fm gets assigned to the new reference
    fm = this.validateAndConvert<FM>(fm, this.FM_SCHEMA, ModelType.FM, opt)

    return fm
  }

  /**
   * Returns *converted value*.
   * Validates (unless `validate=false` passed).
   * Throws only if `throwOnError=true` passed OR if `env().throwOnEntityValidationError`
   */
  validateAndConvert<IN extends BM | DBM | FM, OUT = IN> (
    o: IN,
    schema?: ObjectSchema,
    modelType?: ModelType,
    opt: DaoOptions = {},
  ): OUT {
    // Pre-validation hooks
    if (modelType === ModelType.DBM) {
      o = this.beforeDBMValidate(o as DBM) as IN
    }

    // Return as is if no schema is passed
    if (!schema) {
      return o as any
    }

    // This will Convert and Validate
    const { value, error } = getValidationResult<IN, OUT>(o, schema, this.KIND + (modelType || ''))

    // If we care about validation and there's an error
    if (error && !opt.skipValidation) {
      if (
        (this.baseDatastoreDaoCfg.throwOnEntityValidationError && opt.throwOnError !== false) ||
        opt.throwOnError
      ) {
        if (!error.message) console.log(error)
        throw error
      } else {
        // capture by Sentry and ignore the error
        // It will still *convert* the value and return.
        if (this.baseDatastoreDaoCfg.onError) {
          this.baseDatastoreDaoCfg.onError(error)
        }
      }
    }

    return value // converted value
  }

  create (input: BM, opt: DaoOptions = {}): BM {
    if (opt.throwOnError === undefined) {
      opt.throwOnError = this.baseDatastoreDaoCfg.throwOnDaoCreateObject
    }

    let bm = Object.assign({}, this.beforeCreate(input))
    bm = this.validateAndConvert(bm, this.BM_SCHEMA, ModelType.BM, opt)

    // If no SCHEMA - return as is
    return Object.assign({}, bm)
  }

  /**
   * Will throw Error404 if entity is not found (undefined)
   */
  async requireById (id: string, opt: DaoOptions = {}): Promise<BM> {
    const r = await this.getById(id, opt)
    if (!r) throw new Error(`DB record not found: ${this.KIND}.${id}`)
    return r
  }

  async getByIdAsDBM (id?: string, opt: DaoOptions = {}): Promise<DBM | undefined> {
    const entity = await this.datastoreService.getById(this.KIND, id)
    if (!entity) return

    return this.anyToDBM(entity, opt)
  }

  async getById (id?: string, opt: DaoOptions = {}): Promise<BM | undefined> {
    const entity = await this.datastoreService.getById(this.KIND, id)
    return entity && this.dbmToBM(entity, opt)
  }

  async getByIdAsFM (id?: string, opt: DaoOptions = {}): Promise<FM | undefined> {
    const bm = await this.getById(id, opt)
    return bm && this.bmToFM(bm, opt)
  }

  async getByIdOrCreate (id: string, bmToCreate: BM, opt: DaoOptions = {}): Promise<BM> {
    const bm = await this.getById(id, opt)
    if (bm) return bm

    return this.create(bmToCreate, opt)
  }

  async getByIdOrEmpty (id: string, opt: DaoOptions = {}): Promise<BM> {
    const bm = await this.getById(id, opt)
    if (bm) return bm

    return this.create({ id, ...createdUpdatedFields() } as any, opt)
  }

  /**
   * Saves the object and returns new object (non-mutating) with id/created/updated assigned and validated.
   * todo: refactor to save (bm: BM, opt: DaoSaveOptions = {})
   */
  async save (bm: BM, opt: DaoOptions = {}): Promise<BM> {
    const [savedBM] = await this.saveBatch([bm], opt)
    return savedBM
  }

  async saveAsDBM (dbm: DBM, opt: DaoOptions = {}): Promise<DBM> {
    const [savedDBM] = await this.saveBatchAsDBM([dbm], opt)
    return savedDBM
  }

  async saveBatch (bms: BM[], opt: DaoOptions = {}): Promise<BM[]> {
    const dbms = await this.bmsToDBM(bms, opt)
    const savedDBMs = await this.datastoreService.saveBatch(
      this.KIND,
      dbms,
      this.excludeFromIndexes,
      opt,
    )
    return this.dbmsToBM(savedDBMs, opt)
  }

  async saveBatchAsDBM (_dbms: DBM[], opt: DaoOptions = {}): Promise<DBM[]> {
    const dbms = _dbms.map(dbm => {
      return this.validateAndConvert<DBM>(dbm, this.DBM_SCHEMA!, ModelType.DBM, opt)
    })

    return this.datastoreService.saveBatch(this.KIND, dbms, this.excludeFromIndexes, opt)
  }

  async findBy (by: string, value: any, limit?: number, opt: DaoOptions = {}): Promise<BM[]> {
    const q = this.datastoreService.createQuery(this.KIND).filter(by, value)
    return this.runQuery(q, opt)
  }

  // convenience
  async findByAccountId (accountId: string, opt: DaoOptions = {}): Promise<BM[]> {
    const q = this.datastoreService.createQuery(this.KIND).filter('accountId', accountId)
    return this.runQuery(q, opt)
  }

  async findOneBy (by: string, value: any, opt: DaoOptions = {}): Promise<BM | undefined> {
    const q = this.datastoreService
      .createQuery(this.KIND)
      .filter(by, value)
      .limit(1)
    const bms = await this.runQuery(q, opt)
    return bms.length ? bms[0] : undefined
  }

  async deleteById (id?: string): Promise<void> {
    await this.datastoreService.deleteById(this.KIND, id)
  }

  async deleteBy (by: string, value: any, limit?: number): Promise<void> {
    await this.datastoreService.deleteBy(this.KIND, by, value, limit)
  }

  // convenience
  async deleteByAccountId (accountId: string): Promise<void> {
    await this.datastoreService.deleteBy(this.KIND, 'accountId', accountId)
  }

  createQuery (): Query {
    return this.datastoreService.createQuery(this.KIND)
  }

  /**
   * Careful with this method!:)
   */
  async getAll (opt: DaoOptions = {}): Promise<BM[]> {
    const q = this.datastoreService.createQuery(this.KIND)
    return this.runQuery(q, opt)
  }

  // convenience alias
  getAllQuery (): Query {
    return this.datastoreService.createQuery(this.KIND)
  }

  async runQuery (q: Query, opt: DaoOptions = {}): Promise<BM[]> {
    const entities = await this.datastoreService.runQuery(q)
    return this.dbmsToBM(entities, opt)
  }

  async runQueryAsDBM (q: Query, opt: DaoOptions = {}): Promise<DBM[]> {
    const entities = await this.datastoreService.runQuery(q)
    return this.anyArrayToDBM(entities, opt)
  }

  async runProjectionQuery<T = Partial<DBM>> (q: Query): Promise<T[]> {
    return this.datastoreService.runQuery(q)
  }

  /**
   * Passes data from Datastore "as is" (as DBM)
   */
  streamQueryAsDBM (q: Query, opt: DaoOptions = { skipValidation: true }): Observable<DBM> {
    return this.datastoreService.streamQuery(q).pipe(map(entity => this.anyToDBM(entity, opt)))
  }

  /**
   * Streams as DBM from Datastore and transforms DBM > BM
   */
  streamQueryAsBM (q: Query, opt: DaoOptions = { skipValidation: true }): Observable<BM> {
    return this.datastoreService.streamQuery(q).pipe(flatMap(entity => this.dbmToBM(entity, opt)))
  }

  // proxy method, no added value
  streamQueryIds (q: Query): Observable<string> {
    return this.datastoreService.streamQueryIds(q)
  }

  key (id: string): DatastoreKey {
    return this.datastoreService.key(this.KIND, id)
  }

  // proxy method, no added value
  async countQueryRows (q: Query): Promise<number> {
    return this.datastoreService.countQueryRows(q)
  }

  deserializeJsonField<T = any> (f?: string): T {
    return JSON.parse(f || '{}')
  }

  serializeJsonField (f: any): string | undefined {
    if (f === undefined) return undefined
    return JSON.stringify(f)
  }
}
