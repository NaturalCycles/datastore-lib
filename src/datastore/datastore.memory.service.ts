import { Query } from '@google-cloud/datastore'
import { StringMap } from '@naturalcycles/js-lib'
import { BaseDBEntity, DaoOptions, DatastoreServiceCfg } from './datastore.model'
import { DatastoreService } from './datastore.service'

// Kind > id > entity
type DatastoreData = StringMap<StringMap<any>>

type QueryFilterOperator = '<' | '<=' | '=' | '>=' | '>'

interface QueryFilter {
  name: string
  op: QueryFilterOperator
  val: any
}

type FilterFn = (v: any, val: any) => boolean
const FILTER_FNS: StringMap<FilterFn> = {
  '=': (v, val) => v === val,
  '<': (v, val) => v < val,
  '<=': (v, val) => v <= val,
  '>': (v, val) => v > val,
  '>=': (v, val) => v >= val,
}

/**
 * In-memory limited emulation of DatastoreService API.
 */
export class DatastoreMemoryService extends DatastoreService {
  constructor (datastoreServiceCfg: DatastoreServiceCfg) {
    // super({ project_id: 'Memory' } as any)
    super(datastoreServiceCfg)
  }

  static create (cfg: Partial<DatastoreServiceCfg> = {}): DatastoreMemoryService {
    return new DatastoreMemoryService({
      datastoreOptions: {
        projectId: 'memory',
        credentials: {},
      },
      ...cfg,
    })
  }

  data: DatastoreData = {}

  /**
   * Resets InMemory DB data
   */
  reset (): void {
    this.data = {}
  }

  async getById<T = any> (kind: string, id?: string): Promise<T | undefined> {
    return id && this.data && this.data[kind] && this.data[kind][id]
  }

  async saveBatch<T = any> (
    kind: string,
    _objects: T[],
    excludeFromIndexes?: string[],
    opt: DaoOptions = {},
  ): Promise<(T & BaseDBEntity)[]> {
    const objects = _objects.map(o => this.assignIdCreatedUpdated(o, opt.preserveUpdatedCreated))
    return objects.map(obj => {
      if (!this.data[kind]) this.data[kind] = {}
      this.data[kind][obj.id] = obj

      return obj
    })
  }

  async deleteById (kind: string, id: string): Promise<void> {
    if (!this.data[kind]) this.data[kind] = {}
    delete this.data[kind][id]
  }

  async deleteBy (kind: string, by: string, value: any, limit?: number): Promise<void> {
    this.data[kind] = Object.entries(this.data[kind] || {}).filter(
      ([key, val]) => !(key === by && value === val),
    )
  }

  async runQuery<T = any> (q: Query | any, name?: string): Promise<T[]> {
    // console.log(q)
    const filter: QueryFilter = q.filters.length && q.filters[0]
    // console.log(filter)
    const kind = this.getQueryKind(q)

    let r: any[] = Object.values(this.data[kind] || [])

    if (filter) {
      r = r.filter(v => {
        const fn = FILTER_FNS[filter.op]
        if (fn) return fn(v[filter.name], filter.val)

        console.warn('query filter not supported yet: ', filter)
        return true
      })
    }

    // todo: order, limit

    return r
  }

  async countQueryRows (q: Query): Promise<number> {
    const rows = await this.runQuery(q)
    return rows.length
  }
}
