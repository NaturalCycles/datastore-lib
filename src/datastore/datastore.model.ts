import { DatastoreOptions } from '@google-cloud/datastore'
import { entity } from '@google-cloud/datastore/build/src/entity'
import { JoiValidationError } from '@naturalcycles/nodejs-lib'
import { CredentialBody } from 'google-auth-library'

export type DatastoreKey = entity.Key

export interface DatastorePayload<T = any> {
  key: DatastoreKey
  data: T
  excludeFromIndexes: string[]
}

/**
 * Extends DatastoreOptions by requiring needed properties.
 */
export interface IDatastoreOptions extends DatastoreOptions {
  projectId: string

  /**
   * Object containing client_email and private_key properties
   */
  credentials: CredentialBody
}

export interface DatastoreServiceCfg {
  datastoreOptions: IDatastoreOptions

  /**
   * False will disable all logging (defaul).
   * True will log Datastore operations (but not data).
   */
  log?: boolean

  /**
   * True will log wrote and read entries.
   */
  logData?: boolean

  /**
   * List of Tables to not log data from.
   */
  dontLogTablesData?: string[]
}

export interface BaseDatastoreDaoCfg {
  throwOnEntityValidationError?: boolean
  throwOnDaoCreateObject?: boolean

  /**
   * Called when validation error occurs.
   * Called ONLY when error is NOT thrown (when throwOnEntityValidationError is off)
   */
  onError?: (err: JoiValidationError) => any
}

/**
 * @default All fields default to undefined
 */
export interface DaoOptions {
  skipValidation?: boolean
  throwOnError?: boolean
  preserveUpdatedCreated?: boolean

  /**
   * If true - data will be anonymized (by calling a BaseDao.anonymize() hook that you can extend in your Dao implementation).
   * Only applicable to loading/querying/streaming_loading operations (n/a for saving).
   * There is additional validation applied AFTER Anonymization, so your anonymization implementation should keep the object valid (DBM/BM/FM).
   */
  anonymize?: boolean
}

export interface CreatedUpdated {
  created: number
  updated: number
}

export interface CreatedUpdatedId extends CreatedUpdated {
  id: string
}

export interface CreatedUpdatedVer {
  created: number
  updated: number
  _ver?: number
}

export interface BaseDBEntity {
  id: string
  created: number
  updated: number
  _ver?: number
}

export interface DatastoreStats {
  count: number
}
