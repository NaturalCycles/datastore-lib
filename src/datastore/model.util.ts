import { generateStringId } from '../util/security.util'
import { timeUtil } from '../util/time.util'
import { CreatedUpdated, CreatedUpdatedId } from './datastore.model'

export function createdUpdatedFields (existingObject?: CreatedUpdated): CreatedUpdated {
  const now = timeUtil.nowUnixtime()
  return {
    created: (existingObject && existingObject.created) || now,
    updated: now,
  }
}

export function createdUpdatedIdFields (existingObject?: CreatedUpdatedId): CreatedUpdatedId {
  const now = timeUtil.nowUnixtime()
  return {
    created: (existingObject && existingObject.created) || now,
    id: (existingObject && existingObject.id) || generateStringId(),
    updated: now,
  }
}
