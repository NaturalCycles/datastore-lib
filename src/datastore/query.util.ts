import { Query } from '@google-cloud/datastore'
import { DBQuery } from '@naturalcycles/db-lib'

export function dbQueryToDatastoreQuery (dbQuery: DBQuery, emptyQuery: Query): Query {
  let q = emptyQuery

  // filter
  q = dbQuery._filters.reduce((q, f) => q.filter(f.name, f.op, f.val), q)

  // limit
  q = q.limit(dbQuery._limitValue || 0)

  // order
  q = dbQuery._orders.reduce((q, ord) => q.order(ord.name, { descending: ord.descending }), q)

  // select
  if (dbQuery._selectedFieldNames) {
    q = q.select(dbQuery._selectedFieldNames)
  }

  return q
}
