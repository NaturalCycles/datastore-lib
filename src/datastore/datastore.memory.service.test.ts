import { deepFreeze, mockTime } from '@naturalcycles/test-lib'
import { toArray } from 'rxjs/operators'
import { streamToObservable } from '../util/stream.util'
import { DatastoreMemoryService } from './datastore.memory.service'
import { DatastoreService } from './datastore.service'

interface TestKind {
  id: string
  a: string
  b: string
  c?: string
  d?: string | null
}

const ID = 'randomDatastoreService1'
const KIND = 'TestKind'
let datastoreService: DatastoreService

beforeEach(() => {
  // mocks
  jest.restoreAllMocks()
  mockTime()
  jest.spyOn(require('@naturalcycles/nodejs-lib'), 'stringId').mockImplementation(() => ID)
  datastoreService = DatastoreMemoryService.create({
    log: true,
    logData: true,
  })
})

test('save, generate id, load', async () => {
  const obj = {
    a: 'aa',
    b: 'bb',
    c: 1,
  }
  deepFreeze(obj)

  const savedObj = await datastoreService.save(KIND, obj)
  expect(savedObj).toMatchSnapshot()

  const loaded = await datastoreService.getById(KIND, savedObj.id)
  expect(loaded).toMatchObject(obj)
  expect(loaded).toMatchSnapshot()
})

test('save with id, load, delete', async () => {
  const obj = {
    a: 'aa',
    b: 'bb',
    c: 1,
    id: 'randomid123',
  }
  deepFreeze(obj)

  const savedObj = await datastoreService.save(KIND, obj)
  expect(savedObj.id).toBe(obj.id)
  expect(savedObj).toMatchObject(obj)

  let loaded = await datastoreService.getById(KIND, savedObj.id)
  expect(loaded).toMatchSnapshot()

  await datastoreService.deleteById(KIND, savedObj.id)
  loaded = await datastoreService.getById(KIND, savedObj.id)
  expect(loaded).toBeUndefined()
})

test('saveBatch, findBy, findOneBy, deleteBy', async () => {
  const obj1: any = {
    id: 'id1',
    a: 'aa',
    b: 'b1',
  }
  const obj2: any = {
    id: 'id2',
    a: 'aa',
    b: 'b2',
  }
  const obj3: any = {
    id: 'id3',
    a: 'aa2',
    b: 'b3',
  }
  deepFreeze(obj1)
  deepFreeze(obj2)
  deepFreeze(obj3)

  await datastoreService.saveBatch(KIND, [obj1, obj2, obj3])

  const items = await datastoreService.findBy(KIND, 'a', 'aa')
  expect(items.length).toBe(2)
  expect(items[0]).toMatchObject(obj1)
  expect(items[1]).toMatchObject(obj2)

  const q = datastoreService.createQuery(KIND).filter('a', 'aa')
  const count = await datastoreService.countQueryRows(q)
  // console.log(count)
  expect(count).toBe(2)

  expect(await datastoreService.findOneBy(KIND, 'b', 'b3')).toMatchObject(obj3)
  await datastoreService.deleteBy(KIND, 'b', 'b3')
  expect(await datastoreService.findOneBy(KIND, 'b', 'b3')).toBeUndefined()
})

test('select', async () => {
  const obj1: any = {
    id: 'id1',
    a: 'aa',
    b: 'b1',
  }

  await datastoreService.save(KIND, obj1)

  const q = datastoreService.createQuery(KIND).select('a')
  const rows = await datastoreService.runQuery<TestKind>(q)
  // console.log({rows})
  expect(rows).toEqual([{ a: 'aa' }])
})

function mockTestKindItems (): TestKind[] {
  return [
    {
      id: 'id1',
      a: 'aa',
      b: 'b1',
      c: '2012-12-01',
      d: 'not-null',
    },
    {
      id: 'id2',
      a: 'aa',
      b: 'b2',
      c: '2012-12-02',
      d: 'not-null',
    },
    {
      id: 'id3',
      a: 'aa2',
      b: 'b3',
      c: '2012-12-07',
      d: null,
    },
  ]
}

test('streamQuery', async () => {
  const items = mockTestKindItems()
  await datastoreService.saveBatch(KIND, items)

  const q = datastoreService.createQuery(KIND).filter('a', '=', 'aa')
  const rows = await datastoreService
    .streamQuery<TestKind>(q)
    .pipe(
      // debugging
      // mergeMap(async row => {
      //   console.log({row})
      //   await new Promise(r => setTimeout(r, 500))
      //   return row
      // }, 1),
      // reduce((rows, row) => rows.concat(row), [] as TestKind[]),
      toArray(),
    )
    .toPromise()
  // console.log('done', rows)

  expect(rows.length).toBe(2)
  expect(rows).toMatchSnapshot()

  const ids = await datastoreService
    .streamQueryIds(q)
    .pipe(toArray())
    .toPromise()
  // console.log(ids)
  expect(ids).toEqual(['id1', 'id2'])
})

test('multiple filters', async () => {
  const items = mockTestKindItems()
  await datastoreService.saveBatch(KIND, items)
  const q = datastoreService
    .createQuery(KIND)
    .filter('c', '>=', '2012-12-01')
    .filter('c', '<', '2012-12-03')

  const ids = await datastoreService
    .streamQueryIds(q)
    .pipe(toArray())
    .toPromise()
  // console.log(ids)
  expect(ids).toEqual(['id1', 'id2'])
})

test('null as negation in filters', async () => {
  const items = mockTestKindItems()
  await datastoreService.saveBatch(KIND, items)
  const q = datastoreService.createQuery(KIND).filter('d', '>', null as any)

  const ids1 = await datastoreService
    .streamQueryIds(q)
    .pipe(toArray())
    .toPromise()

  expect(ids1).toEqual(['id1', 'id2'])

  const q2 = datastoreService.createQuery(KIND).filter('d', '<', null as any)

  const ids2 = await datastoreService
    .streamQueryIds(q2)
    .pipe(toArray())
    .toPromise()

  expect(ids2).toEqual(['id1', 'id2'])
})

test('runQueryStream', async () => {
  const items = mockTestKindItems()
  await datastoreService.saveBatch(KIND, items)
  const q = datastoreService.createQuery(KIND).filter('a', '=', 'aa')

  const results = await streamToObservable(datastoreService.runQueryStream(q))
    .pipe(toArray())
    .toPromise()

  expect(results).toMatchSnapshot()
})
