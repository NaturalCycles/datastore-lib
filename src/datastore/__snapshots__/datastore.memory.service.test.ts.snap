// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`save with id, load, delete 1`] = `
Object {
  "a": "aa",
  "b": "bb",
  "c": 1,
  "created": 1529539200,
  "id": "randomid123",
  "updated": 1529539200,
}
`;

exports[`save, generate id, load 1`] = `
Object {
  "a": "aa",
  "b": "bb",
  "c": 1,
  "created": 1529539200,
  "id": "randomDatastoreService1",
  "updated": 1529539200,
}
`;

exports[`save, generate id, load 2`] = `
Object {
  "a": "aa",
  "b": "bb",
  "c": 1,
  "created": 1529539200,
  "id": "randomDatastoreService1",
  "updated": 1529539200,
}
`;
